render/detailed/metrics.go

```diff
@@ -0,0 +1,175 @@
+package detailed
+
+import (
+	"encoding/json"
+	"math"
+
+	"github.com/weaveworks/scope/probe/docker"
+	"github.com/weaveworks/scope/probe/host"
+	"github.com/weaveworks/scope/probe/process"
+	"github.com/weaveworks/scope/report"
+)
+
+const (
+	defaultFormat  = ""
+	filesizeFormat = "filesize"
+	percentFormat  = "percent"
+)
+
+// MetricRow is a tuple of data used to render a metric as a sparkline and
+// accoutrements.
+type MetricRow struct {
+	ID     string
+	Label  string
+	Format string
+	Group  string
+	Value  float64
+	Metric *report.Metric
+}
+
+func (m MetricRow) Copy() MetricRow {
+	metric := m.Metric.Copy()
+	return MetricRow{
+		ID:     m.ID,
+		Label:  m.Label,
+		Format: m.Format,
+		Group:  m.Group,
+		Value:  m.Value,
+		Metric: &metric,
+	}
+}
+
+// MarshalJSON marshals this MetricRow to json. It takes the basic Metric
+// rendering, then adds some row-specific fields.
+func (m MetricRow) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		ID     string  `json:"id"`
+		Label  string  `json:"label"`
+		Format string  `json:"format,omitempty"`
+		Group  string  `json:"group,omitempty"`
+		Value  float64 `json:"value"`
+		report.WireMetrics
+	}{
+		ID:          m.ID,
+		Label:       m.Label,
+		Format:      m.Format,
+		Group:       m.Group,
+		Value:       m.Value,
+		WireMetrics: m.Metric.ToIntermediate(),
+	})
+}
+
+func metricRow(id, label string, metric report.Metric, format, group string) MetricRow {
+	var last float64
+	if s := metric.LastSample(); s != nil {
+		last = s.Value
+	}
+	return MetricRow{
+		ID:     id,
+		Label:  label,
+		Format: format,
+		Group:  group,
+		Value:  toFixed(last, 2),
+		Metric: &metric,
+	}
+}
+
+// toFixed truncates decimals of float64 down to specified precision
+func toFixed(num float64, precision int) float64 {
+	output := math.Pow(10, float64(precision))
+	return float64(int64(num*output)) / output
+}
+
+// NodeMetrics produces a table (to be consumed directly by the UI) based on
+// an origin ID, which is (optimistically) a node ID in one of our topologies.
+func NodeMetrics(n report.Node) []MetricRow {
+	renderers := map[string]func(report.Node) []MetricRow{
+		"process":   processNodeMetrics,
+		"container": containerNodeMetrics,
+		"host":      hostNodeMetrics,
+	}
```

Comment:
I've been giving this some thought, and I think if we move this to a template way of rendering the nodes, and put those templates on the Topology itself (propagated from the Probe), it will allow for plugins to add stuff to the details panel, and make the code in the app truely generic.  In another PR, or course...